Book/php7/extensions_design/hooks.rst

```diff
@@ -0,0 +1,350 @@
+
+Whenever PHP Core or userland code throws an exception the
+``zend_throw_exception_hook`` is called with the exception as argument.
+
+This hooks' signature is fairly simple::
+
+    void my_throw_exception_hook(zval *exception)
+    {
+        if (original_zend_throw_exception_hook != NULL) {
+            original_zend_throw_exception_hook(exception);
+        }
+    }
+
+This hook has no default implementation and points to ``NULL`` if not
+overwritten by an extension.
+
+::
+
+    static void (*original_zend_throw_exception_hook)(zval *ex);
+    void my_throw_exception_hook(zval *exception);
+
+    PHP_MINIT_FUNCTION(my_extension)
+    {
+        original_zend_throw_exception_hook = zend_throw_exception_hook;
+        zend_throw_exception_hook = my_throw_exception_hook;
+
+        return SUCCESS;
+    }
+
+If you implement this hook be aware that this hook is called regardless of
+whether the exception is caught or not. It can still be useful to temporarily
+store the exception here and then combine this with an implementation of the
+Error Handler hook to check if the exception was uncaught and caused the script
+to halt.
+
+Use-cases to implement this hook include debugging, logging and exception
+tracking.
+
+Hooking into eval()
+*******************
+
+PHPs ``eval`` is not an internal function but a special language construct. As
+such you cannot hook into it through ``zend_execute_internal`` or by
+overwriting its function pointer.
+
+Use cases for hooking into eval are not that many, you can use it for profiling
+or for security purposes. If you change its behavior be aware that other extensions
+may need eval. One example is Xdebug that uses it to execute breakpoint conditions.
+
+::
+
+    extern ZEND_API zend_op_array *(*zend_compile_string)(zval *source_string, char *filename);
+
+Hooking into the Garbage Collector
+**********************************
+
+PHPs Garbage Collector can be triggered explicitly when ``gc_collect_cycles()``
+is called or implicitly by the engine itself when the number of collectable
+objects reaches a certain theshold.
+
+To allow understanding of how the garabage collector works or to profile its
+performance, you can overwrite the function pointer hook that performs the
+garbage collection operation. Theoretically you can implement your own garbage
+collection algorithm here, but given other changes to the engine would probably
+be necessary this probably is not really feasible.
+
+::
+
+    int (*original_gc_collect_cycles)(void);
+
+    int my_gc_collect_cycles(void)
+    {
+        original_gc_collect_cycles();
+    }
+
+    PHP_MINIT_FUNCTION(my_extension)
+    {
+        original_gc_collect_cycles = gc_collect_cycles;
+        gc_collect_cycles = my_gc_collect_cycles;
+
+        return SUCCESS;
+    }
+
+Overwrite Interrupt Handler
+***************************
+
+The interrupt handler is called once when the executor global
+``EG(vm_interrupt)`` is set to 1. This is checked after every opcode.  The
```

Comment:
It's not checked after every opcode, just often enough (e.g. on loop backedges) that a check will occur at some point (short of an internal function never returning).